src/driver_camd_midi.c

```diff
@@ -354,55 +329,68 @@ void CamdDrv_MIDI_Unlock() {}
 
 static void /*__saveds*/ __stdargs ServiceTask(void) {
   LOCAL_SYSBASE();
-  LOCAL_REALTIMEBASE();
 
   BOOL initSuccess = FALSE;
 
-  struct Task *thisTask = FindTask(NULL);
+  struct MsgPort *timerMP = NULL;
+  struct timerequest *timerIOReq = NULL;
 
-  if ((g_playerSignalBit = AllocSignal(-1)) == -1) {
+  if ((timerMP = CreateMsgPort()) == NULL) {
     g_error = CamdDrv_Error;
     goto failure;
   }
-
-  ULONG err = 0;
-  g_player = CreatePlayer(
-      PLAYER_Name, (Tag) "JFAudioLib Player", PLAYER_Conductor, (Tag) "JFAudioLib Conductor",
-      PLAYER_AlarmSigTask, (Tag)thisTask, PLAYER_AlarmSigBit,
-      (Tag)g_playerSignalBit, PLAYER_ErrorCode, (Tag)&err, TAG_END);
-
-  if (!g_player) {
+  if ((timerIOReq = (struct timerequest *)CreateIORequest(
+           timerMP, sizeof(struct timerequest))) == NULL) {
     g_error = CamdDrv_Error;
     goto failure;
   }
 
-  SetConductorState(g_player, CONDSTATE_RUNNING, 0);
+  timerIOReq->tr_node.io_Command = TR_ADDREQUEST;
+  timerIOReq->tr_time.tv_secs = 0;
 
-  const ULONG playerSignalBitMask = (1UL << g_playerSignalBit);
-  const ULONG signalMask = playerSignalBitMask | SIGBREAKF_CTRL_C;
-  struct Player *player = g_player;
+  BYTE err = OpenDevice("timer.device", UNIT_ECLOCK, &timerIOReq->tr_node, 37);
+  if (err || timerIOReq->tr_node.io_Device == NULL) {
+    g_error = CamdDrv_Error;
+    goto failure;
+  }
 
-  /*LONG res =*/ SetPlayerAttrs(player, PLAYER_AlarmTime,
-                            player->pl_MetricTime + s_CamdTicksPerMidiFrame,
-                            PLAYER_Ready, TRUE, TAG_END);
+  struct TimerBase *TimerBase = timerIOReq->tr_node.io_Device;
 
   // Let main thread know we're alive
   Signal(g_MainTask, SIGBREAKF_CTRL_E);
 
   initSuccess = TRUE;
 
+  ULONG adjust = 0;
+
+  struct EClockVal eclkStart;
+  ULONG eTicsPerMinute = ReadEClock(&eclkStart) * 60;
+
+  ULONG midiTicsPerMinute = 0;
+  ULONG waitTics = 0;
+
   while (TRUE) {
-    ULONG signals = Wait(signalMask);
-    if ((signals & playerSignalBitMask)) {
+    struct EClockVal eclkStart;
+    ReadEClock(&eclkStart);
+
+    g_service();
 
-      LONG nextAlarm = player->pl_MetricTime + s_CamdTicksPerMidiFrame;
-      // Service the timer function, this is actually parsing the midi song and
-      // causing calls into oiur various FUNC_ midi functions.
-      g_service();
+    struct EClockVal eclkStop;
+    ReadEClock(&eclkStop);
 
-      /*LONG res =*/ SetPlayerAttrs(player, PLAYER_AlarmTime, nextAlarm,
-                                PLAYER_Ready, TRUE, TAG_END);
-    } else if (signals & SIGBREAKF_CTRL_C) {
+    adjust = eclkStop.ev_lo - eclkStart.ev_lo;
+
+    if (midiTicsPerMinute != s_midiTicsPerMinute) {
+      midiTicsPerMinute = s_midiTicsPerMinute;
+      waitTics = eTicsPerMinute / midiTicsPerMinute;
+    }
+
+    timerIOReq->tr_time.tv_micro = waitTics - adjust;
+    DoIO(&timerIOReq->tr_node);
```

Comment:
A (waitTics > adjust) check would be good here to avoid hangs if the delay underflows.

Reply:
Thats a good point! In that case we should probably service the midi routine immediately again.